framework/src/main/java/org/checkerframework/common/basetype/BaseTypeValidator.java

```diff
@@ -112,41 +118,48 @@ protected boolean shouldCheckTopLevelDeclaredType(AnnotatedTypeMirror type, Tree
      *   <li>These properties should also hold recursively for component types of arrays, as wells
      *       as bounds of type variables and wildcards.
      * </ol>
+     *
+     * @param qualifierHierarchy the qualifier hierachy
+     * @param type the type to test
+     * @return list of reasons the type is invalid, or empty list if the type is valid
      */
-    protected Result isValidType(QualifierHierarchy qualifierHierarchy, AnnotatedTypeMirror type) {
-        SimpleAnnotatedTypeScanner<Result, Void> scanner =
-                new SimpleAnnotatedTypeScanner<Result, Void>() {
+    protected List<DiagMessage> isValidType(
+            QualifierHierarchy qualifierHierarchy, AnnotatedTypeMirror type) {
+        SimpleAnnotatedTypeScanner<List<DiagMessage>, Void> scanner =
+                new SimpleAnnotatedTypeScanner<List<DiagMessage>, Void>() {
                     @Override
-                    protected Result defaultAction(AnnotatedTypeMirror type, Void aVoid) {
+                    protected List<DiagMessage> defaultAction(
+                            AnnotatedTypeMirror type, Void aVoid) {
                         return isTopLevelValidType(qualifierHierarchy, type);
                     }
 
                     @Override
-                    protected Result reduce(Result r1, Result r2) {
-                        if (r1 == null) {
-                            if (r2 == null) {
-                                return Result.SUCCESS;
-                            }
-                            return r2;
-                        } else if (r2 == null) {
-                            return r1;
-                        }
-                        return r1.merge(r2);
+                    protected List<DiagMessage> reduce(List<DiagMessage> r1, List<DiagMessage> r2) {
+                        return DiagMessage.mergeLists(r1, r2);
                     }
                 };
         return scanner.visit(type);
     }
 
-    /** Checks every property listed in {@link #isValidType}, but only for the top level type. */
-    protected Result isTopLevelValidType(
+    /**
+     * Checks every property listed in {@link #isValidType}, but only for the top level type. If
+     * successful, returns an empty list. If not successful, returns diagnostics.
+     *
+     * @param qualifierHierarchy the qualifier hierarchy
+     * @param type the type to be checked
+     * @return the diagnostics indicating failure, or an empty list if successful
+     */
+    protected List<DiagMessage> isTopLevelValidType(
```

Comment:
This method either returns a singleton list or the empty list, right?
If so, I would rather model it like `FieldInvariants.isSuperInvariant` and return `DiagMessage` or null.
This makes the code and logic simpler, and the `Result` to `DiagMessage` transition more uniform.

Reply:
Yes, you are right that the method returns a singleton list or the empty list.
Changing its type would make it slightly simpler, but would complicate its caller `isValidType`, which would need to immediately convert the null or DiagMessage into a list.  Take a look and let me know if you agree.